### ISSUE TITLE
Champion: Unmanaged constructed types

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#1937]
* [x] Proposed: https://github.com/dotnet/csharplang/issues/1744
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

**Comment by xoofx** [comment]
@agocke hey, any news on this? We are getting seriously blocked by this at Unity and this is becoming more urgent. How can we help to push for this?

**Comment by xoofx** [comment]
A few questions:

- Is there anymore changes to the spec than what I described [here]?(https://github.com/dotnet/csharplang/issues/1504#issuecomment-430534689)?
- Prototype/Implementation wise, can you confirm that it should be fairly easy to fix this in Roslyn? (apart unlocking the check and adding some tests related to this?)

**Comment by agocke** [comment]
@xoofx This has been approved and @RikkiGibson is almost done with his implementation. Expected release: C# 8.0 😄

**Comment by xoofx** [comment]
> @xoofx This has been approved and @RikkiGibson is almost done with his implementation. Expected release: C# 8.0

Cool! Is it accessible in a branch?

**Comment by agocke** [comment]
@xoofx Just linked to the branch.

**Comment by abelbraaksma** [comment]
@agocke, for reference, maybe it's better to link to the PR directly? I believe it's this one: https://github.com/dotnet/roslyn/pull/31148

**Comment by abelbraaksma** [comment]
@PathogenDavid, thanks, that helps :)

**Comment by hypeartist** [comment]
Hmmm... Seems like the link (https://github.com/dotnet/roslyn/commits/features/unmanaged-constructed-types) is outdated? It gives me 404 error.

**Comment by PathogenDavid** [comment]
[It was merged into the `dev16.1-preview1` branch](https://github.com/dotnet/roslyn/pull/32704).

Based on https://github.com/dotnet/roslyn/issues/31374#issuecomment-459071311, I think development on feature is considered to have been completed.

**Comment by RikkiGibson** [comment]
Sorry, I forgot to check the box for "implementation" :)

Yes, the feature is now in Roslyn's master branch. It won't ship until dev16.1-preview1.

**Comment by dzmitry-lahoda** [comment]
```csharp
    public interface IEntity<TId>
        where TId : unmanaged
    {
        TId Id { get; }
    }
   public interface ISkill : IEntity<(byte, byte)>
   {
   }
```

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS8377	The type '(byte, byte)' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as parameter 'TId' in the generic type or method 'IEntity<TId>'	
```

'ValueTuple<T1,T2>' has no `where T1 : unmanaged where T2 : unmanaged`, but in case presented it seems `unmanaged`. Will(can) it be supported as part of implementation?

**Comment by xoofx** [comment]
@dzmitry-lahoda are you using the Roslyn dev16.1-preview1? (afaik, it is not shipped in any VS2019, you have to fetch the compiler yourself) otherwise I would expect it to work with what did @RikkiGibson

**Comment by xoofx** [comment]
This can issue can be closed right?

**Comment by PathogenDavid** [comment]
Not saying it isn't a little weird, but the champion issues for completed C# features are never closed. (See comments [here](https://github.com/dotnet/csharplang/issues/104#issuecomment-435458830), or [here](https://github.com/dotnet/csharplang/issues/61#issuecomment-292820669) by @gafter)

**Comment by HaloFour** [comment]
@xoofx

IIRC the issue is kept open until the spec and doc work is also complete.

**Comment by xoofx** [comment]
Oh, ok thanks, sorry for the noise 😅
